pyxem/utils/indexation_utils.py

```diff
@@ -923,3 +923,527 @@ def peaks_from_best_vector_match(single_match_result, library, rank=0):
+    polar_image = polar_image/polar_image_norm
+    correlation_array = _match_polar_to_polar_template(polar_image, r, theta, intensity)
+    angle_array = np.arange(correlation_array.shape[0])*delta_theta
+    return angle_array, correlation_array
+
+
+@njit(nogil=True)
+def _match_polar_to_polar_library(polar_image, r_templates, theta_templates, intensities_templates):
+    """Correlates a polar pattern to all polar templates and returns the
+    correlation index and best fitting in-plane angles for each template.
+    Each is returned as a 1D array"""
+    correlations = np.zeros(intensities_templates.shape[0])
+    angles = np.zeros(intensities_templates.shape[0])
+    d_theta = 360 / polar_image.shape[0]
+    for i in prange(intensities_templates.shape[0]):
+        intensities_template = intensities_templates[i]
+        r_template = r_templates[i]
+        theta_template = theta_templates[i]
+        match = _match_polar_to_polar_template(polar_image, r_template, theta_template, intensities_template)
+        correlations[i] = np.max(match)
+        angles[i] = np.argmax(match) * d_theta
+    return correlations, angles
+
+
+@njit
+def _get_correlation_at_angle(polar_image, r_templates, theta_templates, intensities, angle_shifts):
+    """Get the correlation between a polar image and polar templates at particular theta-shifts"""
+    correlations = np.zeros(r_templates.shape[0], dtype=np.float64)
+    for i in range(r_templates.shape[0]):
+        angle = angle_shifts[i]
+        r = r_templates[i]
+        theta = np.mod(theta_templates[i] + angle, polar_image.shape[0])
+        r = r.astype(np.int64)
+        theta = theta.astype(np.int64)
+        intensity = intensities[i]
+        image_intensities = _extract_pixel_intensities(polar_image, r, theta)
+        correlations[i] = np.sum(np.multiply(image_intensities, intensity))
+    return correlations
+    
+    
+def correlate_library_to_pattern(image, simulations, delta_r=1, delta_theta=1, find_direct_beam=True, **kwargs):
+    """
+    Get the best angle and associated correlation values, as well as the correlation with the inverted templates
+    
+    Parameters
+    ----------
+    image : 2d numpy.ndarray
+        The pattern
+    simulations : list of diffsims.sims.diffraction_simulation.DiffractionSimulation
+        The diffraction pattern simulation
+    delta_r : float
+        The sampling of the radial coordinate
+    delta_theta : float
+        The sampling of the azimuthal coordinate
+    find_direct_beam : bool
+        Whether to optimize the direct beam, otherwise the center of the image
+        is chosen
+    **kwargs: passed to the direct beam finding algorithm
+
+    Returns
+    -------
+    angles : 1D numpy.ndarray
+        best fit in-plane angle for each template
+    correlations : 1D numpy.ndarray
+        best correlation for each template
+    correlations_inverse : 1D numpy.ndarray
+        correlation for inverse template of best fit
```

Comment:
Could you explain this to me?

Reply:
When you index a single pattern with "the commercial software", you get a plot of 2 fundamental zones:
![image](https://user-images.githubusercontent.com/25320842/101936601-073c3a00-3be1-11eb-93da-98cb0c9a16a0.png)
I think how it works is one represents the "best match" and the other represents the correlation index of the patterns rotated by 180 degrees (i.e. inverted). The idea is that many diffraction patterns are nearly invertible, they look about the same when rotated by 180 degrees. But most zones don't have real 2-fold rotation symmetry, the effect just arises from Friedl's law. So by comparing the best match and 180 degree match you can see how confidently it indexes the in-plane rotation angle. It's bad when the match and the 180 match are about the same. I wanted to have the option to check this, in my case the result looks like:
![image](https://user-images.githubusercontent.com/25320842/101936057-5766cc80-3be0-11eb-9ed5-0616ae9f5b16.png)
Quite similar, but it's clear that the left hand side has the upper hand. The way I implement it is I plot the maximum over the 360 degrees, then calculate 180 degrees from that maximum. This means that the real solution will always be in the first triangle. The commercial software sometimes has the solution in the second triangle as you can see above. I think this means it looks for a maximum in the range 0-180 and another in 180-360.

Reply:
Ah, it's 180 degree ambiguity stuff, I am glad we have this, from a moving forward perspective it probably just involves cleaning up the explanations